stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/adapter/ChannelListItemAdapter.kt

```diff
@@ -25,45 +34,59 @@ public class ChannelListItemAdapter : BaseChannelListItemAdapter() {
         )
     }
 
-    /**
-     * Returns the layout for the channel items.
-     * Its behavior is such that specifying a layout in the [BaseChannelViewHolderFactory] takes precedence.
-     * If the layout is omitted, the layout specified in the style is used.
-     * If the style layout is omitted, our default layout resource is used.
-     *
-     * @return the resolved layout resource
-     */
-    private fun getChannelItemLayout(): Int =
-        viewHolderFactory.viewHolderLayout
-            ?: style?.channelPreviewLayout
-            ?: R.layout.stream_ui_channel_list_item_view
+    public enum class ChannelItemType {
+        DEFAULT,
+        LOADING_MORE
+    }
+
+    // If we haven't reached the end of the channels, and we're in the last position, we're loading more
+    private fun isLoadingMore(position: Int) = !endReached && position == itemCount - 1
 
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseChannelListItemViewHolder =
-        LayoutInflater.from(parent.context)
-            .inflate(getChannelItemLayout(), parent, false)
-            .let { viewHolderFactory.createChannelViewHolder(it) }
+    override fun getItemViewType(position: Int): Int {
+        return when {
+            isLoadingMore(position) -> ChannelItemType.LOADING_MORE.ordinal
+            else -> ChannelItemType.DEFAULT.ordinal
+        }
+    }
+
+    override fun getItemCount(): Int {
+        return super.getItemCount().let { realCount ->
+            when {
+                // if the list isn't empty, and we haven't reached the end, always offset +1 for the loading more view
+                !endReached && realCount > 0 -> realCount + 1
+                else -> realCount
+            }
+        }
```

Comment:
In my experience, this kind of tricking of RecyclerView will eventually go wrong somewhere when a flag isn't updated or the offset isn't considered in some function. Everywhere else in our SDK we use a real item for the Loading item, that we add to the list that the adapter handles, so that it's rendered through the regular RecyclerView mechanisms.

In this case, we could have something like

```kotlin
sealed class ChannelListItem {
    data class ChannelItem(val channel: Channel) : ChannelListItem()
    object Loading : ChannelListItem()
}
```

Reply:
We could, but is polluting the data set better? I've worked with people on both sides of the fence over the years. Could this even be a decorator? I don't have a strong preference, though. Will rework it if this is blocking.

Reply:
If the implementation works, let's not block things with this now, but we should create a ticket to follow up on it. Fragmentation kills a codebase in the long term.